### PR TITLE
Movie api

### PR DESCRIPTION
**Pull request by Amankarki87**

### CODE REVIEW

File: src/main/java/com/bookMyShow/bookMyShow/Dto/MovieRequestDto.java [hunk 1]
```diff
@@ -1,8 +1,13 @@
...
 
+import com.bookMyShow.bookMyShow.models.Genre;
+import jakarta.validation.constraints.NotEmpty;
 import lombok.Data;
 
 @Data
 public class MovieRequestDto {
+    @NotEmpty(message = "Must not be Empty and NULL")
     private String name;
+    @NotEmptyGenre
```
Comment (priyanshgoel) [inline src/main/java/com/bookMyShow/bookMyShow/Dto/MovieRequestDto.java]: Should reuse @NotNull annotation.

File: src/main/java/com/bookMyShow/bookMyShow/Dto/NotEmptyGenre.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import jakarta.validation.Payload;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Constraint(validatedBy = NotEmptyGenreValidator.class)
+@Target({ ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER })
+@Retention(RetentionPolicy.RUNTIME)
+public @interface NotEmptyGenre {
```
Comment (priyanshgoel) [inline src/main/java/com/bookMyShow/bookMyShow/Dto/NotEmptyGenre.java]: Not required.

File: src/main/java/com/bookMyShow/bookMyShow/Dto/NotEmptyGenreValidator.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.bookMyShow.bookMyShow.Dto;
+
+import com.bookMyShow.bookMyShow.models.Genre;
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+public class NotEmptyGenreValidator implements ConstraintValidator<NotEmptyGenre, Genre> {
```
Comment (priyanshgoel) [inline src/main/java/com/bookMyShow/bookMyShow/Dto/NotEmptyGenreValidator.java]: Not required.

File: src/main/java/com/bookMyShow/bookMyShow/controllers/MovieController.java [hunk 4]
```diff
@@ -18,8 +19,11 @@ public class MovieController {
     MovieService movieService;
```
Comment (priyanshgoel) [inline src/main/java/com/bookMyShow/bookMyShow/controllers/MovieController.java]: private

File: src/main/java/com/bookMyShow/bookMyShow/controllers/MovieController.java [hunk 5]
```diff
@@ -18,8 +19,11 @@ public class MovieController {
...
 
     @PostMapping
-    public ResponseEntity<ResponseFormatDto> save(@RequestBody MovieRequestDto movieRequestDto) {
-        Movie movie = movieService.save(movieRequestDto.getName());
+    public ResponseEntity<ResponseFormatDto> save(@Valid @RequestBody MovieRequestDto movieRequestDto) {
```
Comment (priyanshgoel) [inline src/main/java/com/bookMyShow/bookMyShow/controllers/MovieController.java]: save is a DB operation. the naming should be createMovie()

File: src/main/java/com/bookMyShow/bookMyShow/controllers/MovieController.java [hunk 6]
```diff
@@ -18,8 +19,11 @@ public class MovieController {
...
+        Movie movie = movieService.save(
```
Comment (priyanshgoel) [inline src/main/java/com/bookMyShow/bookMyShow/controllers/MovieController.java]: Same as above.

File: src/main/java/com/bookMyShow/bookMyShow/Dto/ResponseFormatDto.java [hunk 7]
```diff
@@ -14,5 +14,5 @@ public class ResponseFormatDto<T> {
     private HttpStatus statusCode;
     private String message;
     private T data;
-    private Map<String,T> meta = new HashMap<>(); // Use for holding pagination or any other additional properties for the payload.
+    private Map<String,T> meta = new HashMap<String,T>(); // Use for holding pagination or any other additional properties for the payload.
```
Comment (priyanshgoel) [inline src/main/java/com/bookMyShow/bookMyShow/Dto/ResponseFormatDto.java:14]: ResponseFormatDTO - > may or may not be generic. Also - Doesn't need status, statuscode, etc.

File: src/main/java/com/bookMyShow/bookMyShow/services/AuditoriumServiceImpl.java [hunk 8]
```diff
@@ -30,16 +30,13 @@ public Auditorium save(String name,Long theatreId) {
         Optional<Theatre> theatre = theatreRepository.findById(theatreId);
 
         if(!theatre.isPresent()) {
-            ErrorResponse errorResponse = ErrorResponse.builder()
+
+            Error error = Error.builder()
```
Comment (priyanshgoel) [inline src/main/java/com/bookMyShow/bookMyShow/services/AuditoriumServiceImpl.java:34]: Error => ElementNotFoundException.

File: src/main/java/com/bookMyShow/bookMyShow/services/MovieServiceImpl.java [hunk 9]
```diff
@@ -1,20 +1,40 @@
...
 
     @Override
-    public Movie save(String name) {
-        Movie movie = new Movie();
-        movie.setName(name);
-        Movie result = movieRepository.save(movie);
+    public Movie save(String name,Genre movieGenre) {
+        Optional<Movie> movie = movieRepository.findByName(name);
+
+        if (movie.isPresent()) {
```
Comment (priyanshgoel) [inline src/main/java/com/bookMyShow/bookMyShow/services/MovieServiceImpl.java:21]: Same as above.

File: src/main/java/com/bookMyShow/bookMyShow/services/MovieServiceImpl.java [hunk 10]
```diff
@@ -1,20 +1,40 @@
...
+            Error error = Error.builder()
+                    .code(HttpStatus.CONFLICT)
+                    .message(ApiConstant.MOVIE_ALREADY_EXISTS)
+                    .status(ApiConstant.ERROR)
+                    .build();
+            throw error;
+        }
+
+        Movie moviedto = Movie
```
Comment (priyanshgoel) [inline src/main/java/com/bookMyShow/bookMyShow/services/MovieServiceImpl.java]: movieDTO

File: src/test/java/bookMyShow/services/MovieServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        when(movieRepository.save(moviedto)).thenReturn(moviedto);
+        Movie result = movieService.save("Test",Genre.COMEDY);
+        Assert.assertEquals(result.getName(), "Test");
+        Assert.assertEquals(result.getMovieGenre(),Genre.COMEDY);
+    }
+
+
+    @Test
+    public void saveMovieTestWhenConflict() {
+        Movie movie = Movie
```
Comment (priyanshgoel) [inline src/test/java/bookMyShow/services/MovieServiceTest.java]: move to utility/global.

File: src/main/java/com/bookMyShow/bookMyShow/controllers/MovieController.java (hunk 4 above)
Comment (Amankarki87) [reply src/main/java/com/bookMyShow/bookMyShow/controllers/MovieController.java]: Resolved

File: src/main/java/com/bookMyShow/bookMyShow/controllers/MovieController.java (hunk 5 above)
Comment (Amankarki87) [reply src/main/java/com/bookMyShow/bookMyShow/controllers/MovieController.java]: Remove

File: src/main/java/com/bookMyShow/bookMyShow/controllers/MovieController.java (hunk 6 above)
Comment (Amankarki87) [reply src/main/java/com/bookMyShow/bookMyShow/controllers/MovieController.java]: Resolved

File: src/main/java/com/bookMyShow/bookMyShow/Dto/ResponseFormatDto.java (hunk 7 above)
Comment (Amankarki87) [reply src/main/java/com/bookMyShow/bookMyShow/Dto/ResponseFormatDto.java:14]: Removed statusCode,message and status from ResponseFormatDto.

File: src/main/java/com/bookMyShow/bookMyShow/services/MovieServiceImpl.java (hunk 10 above)
Comment (Amankarki87) [reply src/main/java/com/bookMyShow/bookMyShow/services/MovieServiceImpl.java]: Resolved

File: src/main/java/com/bookMyShow/bookMyShow/services/MovieServiceImpl.java (hunk 9 above)
Comment (Amankarki87) [reply src/main/java/com/bookMyShow/bookMyShow/services/MovieServiceImpl.java:21]: Resolved

File: src/main/java/com/bookMyShow/bookMyShow/services/AuditoriumServiceImpl.java (hunk 8 above)
Comment (Amankarki87) [reply src/main/java/com/bookMyShow/bookMyShow/services/AuditoriumServiceImpl.java:34]: Resolved

File: src/main/java/com/bookMyShow/bookMyShow/Dto/NotEmptyGenreValidator.java (hunk 3 above)
Comment (Amankarki87) [reply src/main/java/com/bookMyShow/bookMyShow/Dto/NotEmptyGenreValidator.java]: Resolved

File: src/main/java/com/bookMyShow/bookMyShow/Dto/NotEmptyGenre.java (hunk 2 above)
Comment (Amankarki87) [reply src/main/java/com/bookMyShow/bookMyShow/Dto/NotEmptyGenre.java]: Resolved

File: src/test/java/bookMyShow/services/MovieServiceTest.java (hunk 11 above)
Comment (Amankarki87) [reply src/test/java/bookMyShow/services/MovieServiceTest.java]: Resolved

File: src/main/java/com/bookMyShow/bookMyShow/Dto/MovieRequestDto.java (hunk 1 above)
Comment (Amankarki87) [reply src/main/java/com/bookMyShow/bookMyShow/Dto/MovieRequestDto.java]: Resolved

**Pull request merged**